lib/MetricsProducer.js

```diff
@@ -0,0 +1,63 @@
+'use strict'; // eslint-disable-line strict
+
+const { Logger } = require('werelogs');
+
+const BackbeatProducer = require('./BackbeatProducer');
+const MetricsModel = require('./models/MetricsModel');
+
+class MetricsProducer {
+    /**
+    * @constructor
+    * @param {object} zkConfig - zookeeper configurations
+    * @param {object} mConfig - metrics configurations
+    */
+    constructor(zkConfig, mConfig) {
+        this._zkConfig = zkConfig;
+        this._topic = mConfig.topic;
+
+        this._producer = null;
+        this._log = new Logger('MetricsProducer');
+    }
+
+    getBatch() {
+        return this.batch;
+    }
+
+    setupProducer(done) {
+        const producer = new BackbeatProducer({
+            zookeeper: { connectionString: this._zkConfig.connectionString },
+            topic: this._topic,
+        });
+        producer.once('error', done);
+        producer.once('ready', () => {
+            producer.removeAllListeners('error');
+            producer.on('error', err => {
+                this._log.error('error from backbeat producer',
+                               { error: err });
+            });
+            this._producer = producer;
+            done();
+        });
+    }
+
+    /**
+     * @param {object} extMetrics - an object where keys are all buckets for a
+     *   given extension and values are the metrics for the bucket
+     *   (i.e. { my-bucket: { ops: 1, bytes: 124 } } )
+     * @param {string} type - type of metric (queueud or processed)
+     * @param {string} ext - extension (i.e. 'crr')
+     * @param {function} cb - callback
+     * @return {undefined}
+     */
+    publishMetric(extMetrics, type, ext, cb) {
+        Object.keys(extMetrics).forEach(bucketName => {
```

Comment:
`forEach` is syncronous, when you would want to wait until all extensions were published before calling the initial callback.

Reply:
The `cb` is refactored out now and instead, if the producer returns an error, `log.trace` but continue the forEach loop :) Thanks!